zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala

```diff
@@ -0,0 +1,115 @@
+package zio.crypto.hashing
+
+import zio._
+import zio.crypto.ByteHelpers
+
+import java.nio.charset.Charset
+import java.security.{ MessageDigest => JMessageDigest }
+
+sealed class HashAlgorithm(val name: String)
+
+object HashAlgorithm {
+  case object SHA256 extends HashAlgorithm("SHA-256")
+  case object SHA512 extends HashAlgorithm("SHA-512")
+
+  // DO NOT USE IN A SECURE CONTEXT
+  case object MD5 extends HashAlgorithm("MD5")
```

Comment:
Let's solve this in a different way:

```scala
trait Secure[-Algorithm]
```

Then inside HashAlgorithm companion object:

```scala
implicit val secureSha256: Secure[SHA256.type] = new Secure[SHA256.type]{}
...
```

Then make a helper in `zio.crypto`:

```scala
def unsecure[A](f: Secure[Any] => A): A = f(new Unsecure[Any]{})
```

Then any code which requires a `HashAlgorithm` should also require `Secure` for that hash algorithm.

e.g.:

```scala
def hash[Alg <: HashAlgorithm](m: Array[Byte], alg: HashAlgorithm)(implicit s: Secure[HashAlgorihm]): Task[MessageDigest[Array[Byte]]]
```

Now if you want to use a secure algorithm, it will be available. But if you want to use an insecure algorithm, you have to do:

```scala
unsecure { implicit s =>
 ...
}
```

The technique can be used for anything, even if it's not a hash.